### PR TITLE
Add initial interface for Data comparison framework

### PR DESCRIPTION
**Pull request by bishwajit-db**
Add initial interface for Data comparison framework

### CODE REVIEW

**Comment by sundarshankar89** [comment]
@nkvuong Initial WIP PR.

File: pyproject.toml [hunk 1]
```diff
@@ -76,6 +76,7 @@ dependencies = [
     "ruff~=0.3.4",
     "types-PyYAML~=6.0.12",
     "types-requests~=2.31.0",
+    "pyspark>=3.5.0",
```
[the remark this answers is not in the document]
Comment (bishwajit-db) [reply pyproject.toml]: Removed

File: src/databricks/labs/ucx/recon/_schema_comparator.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+from pyspark.sql import DataFrame
+from pyspark.sql.types import StructField, StructType, DataType, ArrayType
+
+from databricks.labs.ucx.recon import SchemaComparator, SchemaComparisonEntry, SchemaComparisonResult
+
+
+class StandardSchemaComparator(SchemaComparator):
+
+    def compare_schema(self, source: DataFrame, target: DataFrame) -> SchemaComparisonResult:
```
[the remark this answers is not in the document]
Comment (bishwajit-db) [reply src/databricks/labs/ucx/recon/_schema_comparator.py]: Done

File: src/databricks/labs/ucx/recon/_base.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    is_matching: bool
+    data: list[SchemaComparisonEntry]
+
+
+@dataclass
+class DataComparisonResult:
+    pass
+
+
+class TableMetadataRetriever(ABC):
```
Comment (sundarshankar89) [inline src/databricks/labs/ucx/recon/_base.py:57]: do you really want this to be separate, not as a property of the profiler?

File: src/databricks/labs/ucx/recon/_data_profiler.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from pyspark.sql import DataFrame
+
+from databricks.labs.ucx.recon import DataProfiler, DataProfilingResult
+
+
+class StandardDataProfiler(DataProfiler):
+    def profile_data(self, source: DataFrame) -> DataProfilingResult:
```
[the remark this answers is not in the document]
Comment (bishwajit-db) [reply src/databricks/labs/ucx/recon/_data_profiler.py]: Done

File: src/databricks/labs/ucx/recon/_data_profiler.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+from databricks.labs.ucx.recon import DataProfiler, DataProfilingResult, TableDescriptor, TableMetadataRetriever
+
+
+class StandardDataProfiler(DataProfiler):
+    def __init__(self, sql_backend: SqlBackend, metadata_retriever: TableMetadataRetriever):
+        self._sql_backend = sql_backend
+        self._metadata_retriever = metadata_retriever
+
+    def profile_data(self, source: TableDescriptor) -> DataProfilingResult:
```
Comment (sundarshankar89) [inline src/databricks/labs/ucx/recon/_data_profiler.py:14]: nit:
```suggestion
    def profile_data(self, entity: TableDescriptor) -> DataProfilingResult:
```

File: src/databricks/labs/ucx/recon/_data_profiler.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        row_count = self._get_table_row_count(source)
+        return DataProfilingResult(
+            row_count,
+            self._metadata_retriever.get_metadata(source),
+        )
+
+    def _get_table_row_count(self, source: TableDescriptor) -> int:
```
Comment (sundarshankar89) [inline src/databricks/labs/ucx/recon/_data_profiler.py:21]: nit:
```suggestion
    def _get_table_row_count(self, entity: TableDescriptor) -> int:
```

File: src/databricks/labs/ucx/recon/_metadata_retriever.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        query_result: Iterator[Row] = self._sql_backend.fetch(schema_query)
+        columns = {
+            ColumnMetadata(str(row["col_name"]), str(row["data_type"]))
+            for row in query_result
+            if not str(row["col_name"]).startswith("#")
+        }
+        return TableMetadata(source, list(columns))
+
+
+def _prepare_metadata_query(source: TableDescriptor) -> str:
```
Comment (sundarshankar89) [inline src/databricks/labs/ucx/recon/_metadata_retriever.py:26]: nit: 
```suggestion
def _prepare_metadata_query(entity: TableDescriptor) -> str:
```

File: src/databricks/labs/ucx/recon/_metadata_retriever.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+from databricks.labs.ucx.recon import TableDescriptor, TableMetadata, ColumnMetadata
+from databricks.labs.ucx.recon import TableMetadataRetriever
+
+
+class DatabricksTableMetadataRetriever(TableMetadataRetriever):
+    def __init__(self, sql_backend: SqlBackend):
+        self._sql_backend = sql_backend
+
+    def get_metadata(self, source: TableDescriptor) -> TableMetadata:
```
Comment (sundarshankar89) [inline src/databricks/labs/ucx/recon/_metadata_retriever.py:15]: nit:
```suggestion
    def get_metadata(self, entity: TableDescriptor) -> TableMetadata:
```

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/recon/_base.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+
+@dataclass
+class TableDescriptor:
```
Comment (nkvuong) [inline src/databricks/labs/ucx/recon/_base.py:6]: we already have a Table dataclass that you can reuse
https://github.com/databrickslabs/ucx/blob/953c4837614b355d03bac9c0f201ca2a39e11a14/src/databricks/labs/ucx/hive_metastore/tables.py#L50-L54

File: src/databricks/labs/ucx/recon/_base.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+
+@dataclass
+class DataProfilingResult:
+    row_count: int
+    table_metadata: TableMetadata
+
+
+@dataclass
+class SchemaComparisonEntry:
```
Comment (nkvuong) [inline src/databricks/labs/ucx/recon/_base.py:37]: you could reuse/extend https://github.com/databrickslabs/ucx/blob/953c4837614b355d03bac9c0f201ca2a39e11a14/src/databricks/labs/ucx/hive_metastore/mapping.py#L23-L29
or 
https://github.com/databrickslabs/ucx/blob/953c4837614b355d03bac9c0f201ca2a39e11a14/src/databricks/labs/ucx/hive_metastore/mapping.py#L63-L65

File: src/databricks/labs/ucx/recon/_schema_comparator.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        source_column=source_col.name if source_col else None,
+        source_datatype=source_col.data_type if source_col else None,
+        target_column=target_col.name if target_col else None,
+        target_datatype=target_col.data_type if target_col else None,
+        is_matching=is_matching,
+        notes=notes,
+    )
+
+
+def _compare_col_metadata(source_col: ColumnMetadata, target_col: ColumnMetadata) -> bool:
```
Comment (nkvuong) [inline src/databricks/labs/ucx/recon/_schema_comparator.py:54]: should define a `__eq__` method for `ColumnMetadata` class instead

**Review by nkvuong — CHANGES_REQUESTED** [review]
don't forget to run `make fmt`

also since this decreases code coverage, it won't be auto-merged :)

**Review by nkvuong — APPROVED** [review]

**Review by sundarshankar89 — APPROVED** [review]
LGTM

**Pull request merged**